### PR TITLE
LOK-2516 - Implementation of WSMAN Monitor

### PR DESCRIPTION
**Pull request by naeem-afzal**
## Description
Bring existing functionality from Horizon with minimum possible changes. 

Use existing wsman library from Horizon: https://github.com/OpenNMS/wsman

## Jira link(s)
- https://opennms.atlassian.net/browse/LOK-2516

## Checklist
* [ ] Follows Lōkahi's [development guidelines.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Development-Guidelines)
* [ ] Appropriate reviewer(s) have been selected.
* [ ] Jira issue(s) have been updated to "In Review".
* [ ] Includes [appropriate tests.](https://github.com/OpenNMS-Cloud/lokahi/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.
* [ ] Notify devops of changes to the Charts.
* [ ] Notify documentation team of any changes to names of screens or features (affects URLs).

### CODE REVIEW

File: minion/plugins/wsman-plugin/src/main/feature/feature.xml [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        <!-- false to avoid restarts of proxy which cascades badly -->
+        <bundle dependency="false" start-level="20">mvn:org.ow2.asm/asm/9.5</bundle>
+        <bundle dependency="false" start-level="20">mvn:org.ow2.asm/asm-util/9.5</bundle>
+        <bundle dependency="false" start-level="20">mvn:org.ow2.asm/asm-tree/9.5</bundle>
+        <bundle dependency="false" start-level="20">mvn:org.ow2.asm/asm-analysis/9.5</bundle>
+        <bundle dependency="false" start-level="20">mvn:org.ow2.asm/asm-commons/9.5</bundle>
+        <bundle start-level="20">mvn:org.apache.aries.proxy/org.apache.aries.proxy/1.1.13</bundle>
+    </feature>
+
+    <feature name="aries-blueprint" description="Aries Blueprint" version="4.3.10">
```
Comment (cgorantla) [inline minion/plugins/wsman-plugin/src/main/feature/feature.xml:135]: we don't need to define `aries-blueprint` separately, it's should be already present in karaf.

**Review by cgorantla — COMMENTED** [review]
Looks good overall, few comments

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanDetector.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                WSManClient client = m_factory.getClient(endpoint);
+
+                Identity identity = null;
+                final Map<String, String> attributes = new HashMap<>();
+                try {
+                    identity = client.identify();
+
+                } catch (WSManException e) {
+                    LOG.info("Identify failed for endpoint {} with error {}.", endpoint, e);
+                    System.out.println(">>> WSManException: " + e.toString());
```
Comment (cgorantla) [inline minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanDetector.java:105]: remove sout

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanDetector.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                }
+
+                if (identity != null) {
+                    serviceStatus = true;
+                    LOG.info("\n\n>> UP, Node is UP, Congrats, WSMAN identified/ detected successfully. {} ", endpoint);
+                } else LOG.info(">> Node DOWN, not able to detect {} ", endpoint);
```
Comment (cgorantla) [inline minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanDetector.java:111]:
```suggestion
                } else {
                    LOG.info(" Node DOWN, not able to detect {} ", endpoint);
                }
```
remove >> in the logs.

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanDetector.java [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+                                .setIpAddress(serviceResult.getIpAddress())
+                                .setService(serviceResult.getService())
+                                .build();
+                        return serviceResponse;
+                    });
+
+            return future;
+
+        } catch (IllegalArgumentException e) {
+            LOG.debug("Invalid WSMAN Criteria during detection of interface {}", host, e);
```
Comment (cgorantla) [inline minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanDetector.java:74]: remove sout

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+            }
+
+            // Now you can access and print all the fields of the wsmanRequest object
+            LOG.debug("\n\n-----------New WSMAN Request-------------");
+            LOG.debug("Host: {}", wsmanConfiguration.getHost());
+            LOG.debug("Port: {}", wsmanConfiguration.getPort());
+            LOG.debug("Path: {}", wsmanConfiguration.getPath());
+            LOG.debug("Username: {}", wsmanConfiguration.getUsername());
+            LOG.debug("Password: {}", wsmanConfiguration.getPassword());
+            LOG.debug("GSS Auth: {}", wsmanConfiguration.getGssAuth());
```
Comment (cgorantla) [inline minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java:109]: remove this from log

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+            LOG.debug("POLL - WsManMonitor returned normally.");
+            return future;
+
+        } catch (Exception e) {
+            LOG.debug(
+                    "Unexpected exception during WSMAN poll {}, {}, {}",
+                    wsmanMonitorRequest.getAgentConfiguration().getHost(),
+                    wsmanMonitorRequest.getAgentConfiguration().getPort(),
+                    e.getMessage());
```
Comment (cgorantla) [inline minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java:169]:
```suggestion
                    e)
```

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+                } catch (Exception e) {
+                    LOG.debug(">> WSMAN - Node DOWN: {}, {} ", wsmanConfiguration.getHost(), e.toString());
+                    ps.setReason("Exception: " + e.toString());
+                }
+
+            } catch (Exception e) {
+                LOG.debug(
+                        ">> WSMAN - Node DOWN Hay: {}, {} ",
+                        wsmanMonitorRequest.getAgentConfiguration().getHost(),
+                        e.toString());
```
Comment (cgorantla) [inline minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java:248]: adding exception e at the should be enough, no need for .toString()

File: minion/plugins/shell-commands/src/main/java/org/opennms/horizon/minion/icmp/shell/WSManDetectorCommand.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    // -p "wsmanuser4pwd" --node-id 1110 --monitor-service-id 2220 -v 0 -rt 1
+    //    wsman-detector -m 9999 -s 1000 -t 5000 -r 5000 -host "win10.com" -port "5985" -path "/wsman" -u "wsmanuser" -p
+    // "wsmanuser4pwd" --node-id 1110 --monitor-service-id 2220 -v 0 -rt 1
+
+    @Option(aliases = "--node-id", name = "-nid", description = "Node id")
+    Long nodeId = 1111L;
+
+    @Option(aliases = "--monitor-service-id", name = "-msid", description = "Monitored service id")
+    Long monServiceId = 2222L;
+
```
[the remark this answers is not in the document]
Comment (naeem-afzal) [reply minion/plugins/shell-commands/src/main/java/org/opennms/horizon/minion/icmp/shell/WSManDetectorCommand.java]: random values has been removed

File: minion/plugins/wsman-plugin/src/main/feature/feature.xml (hunk 1 above)
Comment (naeem-afzal) [reply minion/plugins/wsman-plugin/src/main/feature/feature.xml]: aries-blueprint sections has been removed

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanDetector.java (hunk 2 above)
Comment (naeem-afzal) [reply minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanDetector.java]: System.out.println has been removed

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanDetector.java (hunk 3 above)
Comment (naeem-afzal) [reply minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanDetector.java]: >> signs has been removed

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java (hunk 5 above)
Comment (naeem-afzal) [reply minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java]: removed

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java (hunk 6 above)
Comment (naeem-afzal) [reply minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java]: done

File: minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java (hunk 7 above)
Comment (naeem-afzal) [reply minion/plugins/wsman-plugin/src/main/java/org/opennms/lokahi/minion/wsman/WsmanMonitor.java]: toString removed

**Review by cgorantla — APPROVED** [review]
LGTM!

**Pull request merged**